ta4j-core/src/main/java/org/ta4j/core/indicators/statistics/VarianceIndicator.java

```diff
@@ -25,53 +25,122 @@
 
 import org.ta4j.core.Indicator;
 import org.ta4j.core.indicators.CachedIndicator;
-import org.ta4j.core.indicators.SMAIndicator;
 import org.ta4j.core.num.Num;
+import org.ta4j.core.num.NumFactory;
 
 /**
- * Variance indicator.
+ * Variance indicator with optimized calculation for sequential access.
+ * Uses Welford's online algorithm for sequential variance calculation.
  */
-public class VarianceIndicator extends CachedIndicator<Num> {
+public final class VarianceIndicator extends CachedIndicator<Num> {
 
     private final Indicator<Num> indicator;
     private final int barCount;
-    private final SMAIndicator sma;
-
-    /**
-     * Constructor.
-     *
-     * @param indicator the indicator
-     * @param barCount  the time frame
-     */
-    public VarianceIndicator(Indicator<Num> indicator, int barCount) {
+    private final boolean isSampleVariance;
+
+    // State for sequential calculation
+    private int previousIndex = -1;
+    private Num previousSum;
```

Comment:
There was an [effort to remove class instance variables from cached indicators](https://github.com/ta4j/ta4j/issues/906) so this would be a regression. Improving performance will require a global approach.

Reply:
What kind of concurrency are we talking about? 

It is more efficient to scale vertically than horizontally. Adding CPUs to slow algorithm makes it "maybe less slow" but "more expensive" if you pay CPU in the cloud.

I can imagine following scenarios:

- Divide and Conquer per indicator - this may have large overhead
- Divide and Conquer per strategy - already implemented in BacktestExecutor and it is compatible with fast path

Reply:
You probably meant sharing of indicator between multiple strategies. Hm. It is tricky, because calculation is stateful.

I can imagine following approaches:

- Run Strategy per thread  with separate indicator contexts. It is easy to do, but each timepoint is calculated separately and this wastes CPU time.
- Run single threaded simulation of multiple strategies sharing single indicator context. This does not waste CPU, nor gains multithreading performance. It is small optimization before parallelization.
- **Two phase run. First run fills cache without strategy testing. Second run can execute thread per strategy (thread count = CPU count).** Currently I use only statefulness of each indicator without cache over each timepoint. Two phase approach may be the best for bar simulations as it prepares state without contention, and then all strategies just read prepared states. This behavior is similar to current `CachedIndicator`

IndicatorContext is constraint for threading as it is bound to certain time point.